lib/ssl/src/tls_connection_1_3.erl

```diff
@@ -446,13 +445,15 @@ wait_sh(internal, #server_hello{} = Hello,
              tls_gen_connection:next_event(wait_ee, no_record, State1)
     end;
 wait_sh(internal, #server_hello{} = Hello, 
-         #state{ssl_options = #{middlebox_comp_mode := true}, protocol_specific = PS} = State0) ->
+        #state{protocol_specific = PS} = State0) ->
     IsRetry = maps:get(hello_retry, PS, false),
     case tls_handshake_1_3:do_wait_sh(Hello, State0) of
         #alert{} = Alert ->
             ssl_gen_statem:handle_own_alert(Alert, wait_sh, State0);
-         {State1, start, ServerHello} ->
-            %% hello_retry_request: go to start
+        {State1 = #state{session = #session{session_id = Id}}, start, ServerHello} when Id =/= ?EMPTY_ID ->
+            %% hello_retry_request
+            {next_state, hello_retry_middlebox_assert, State1, [{next_event, internal, ServerHello}]};
+        {State1 = #state{session = #session{session_id = Id}}, start, ServerHello} when Id == ?EMPTY_ID ->
             {next_state, hello_retry_middlebox_assert, State1, [{next_event, internal, ServerHello}]};
```

Comment:
just checking: return values on lines 455, 457 are the same - is this intended?

currently 2 patterns from line 453, 456 could be collapsed into single one ... you have a code comment anyway... maybe it could be extended to decribe that no matter what Id value is the action is the same ...

Reply:
Good catch, one of those clauses where dead code. I made some clarifications to the code.